oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java

```diff
@@ -0,0 +1,296 @@
+        }
+        return updatesMap.containsValue(Boolean.TRUE);
+    }
+
+    private class OakBulkProcessorListener implements BulkProcessor.Listener {
+
+        @Override
+        public void beforeBulk(long executionId, BulkRequest bulkRequest) {
+            // register new bulk party
+            phaser.register();
+
+            // init update status
+            updatesMap.put(executionId, Boolean.FALSE);
+
+            LOG.debug("Sending bulk with id {} -> {}", executionId, bulkRequest.getDescription());
+            if (LOG.isTraceEnabled()) {
+                LOG.trace("Bulk Requests: \n{}", bulkRequest.requests()
+                        .stream()
+                        .map(DocWriteRequest::toString)
+                        .collect(Collectors.joining("\n"))
+                );
+            }
+        }
+
+        @Override
+        public void afterBulk(long executionId, BulkRequest bulkRequest, BulkResponse bulkResponse) {
+            LOG.debug("Bulk with id {} processed with status {} in {}", executionId, bulkResponse.status(), bulkResponse.getTook());
+            if (LOG.isTraceEnabled()) {
+                try {
+                    LOG.trace(Strings.toString(bulkResponse.toXContent(jsonBuilder(), EMPTY_PARAMS)));
+                } catch (IOException e) {
+                    LOG.error("Error decoding bulk response", e);
+                }
+            }
+            if (bulkResponse.hasFailures()) { // check if some operations failed to execute
+                NodeBuilder status = definitionBuilder.child(IndexDefinition.STATUS_NODE);
+                // Read the current failed paths (if any) on the :status node into failedDocList
+                if (status.hasProperty(IndexDefinition.FAILED_DOC_PATHS)) {
+                    for (String str : status.getProperty(IndexDefinition.FAILED_DOC_PATHS).getValue(Type.STRINGS)) {
+                        failedDocSet.add(str);
+                    }
+                }
+
+                for (BulkItemResponse bulkItemResponse : bulkResponse) {
+                    if (bulkItemResponse.isFailed()) {
+                        BulkItemResponse.Failure failure = bulkItemResponse.getFailure();
+                        failedDocSet.add(bulkItemResponse.getId());
+                        // Log entry to be used to parse logs to get the failed doc id/path if needed
+                        LOG.error("ElasticIndex Update Doc Failure: Error while adding/updating doc with id : [{}]", bulkItemResponse.getId());
+                        LOG.error("Failure Details: BulkItem ID: " + failure.getId() + ", Failure Cause: {}", failure.getCause());
+                    } else {
+                        // Set indexUpdated to true even if 1 item was updated successfully
+                        updatesMap.put(executionId, Boolean.TRUE);
+                    }
+                }
+
+                if (failedDocSet.size() > FAILED_DOC_COUNT_FOR_STATUS_NODE) {
```

Comment:
The problem with this approach is: let's say there is one batch that has more than FAILED_DOC_COUNT_FOR_STATUS_NODE failed documents. In that case, errors and an info message is logged, but the property FAILED_DOC_PATHS isn't set. That means, in order to check if there are any failed docs, you need to monitor the log file.

What about, in case there are more than FAILED_DOC_COUNT_FOR_STATUS_NODE failed documents, remove those that are too many (that is, truncate the size). That way, a monitoring system could just monitor if there are any property IndexDefinition.FAILED_DOC_PATHS set. And checking the log file is only needed to find the paths.

Reply:
I have changed the logic so we never have more failed docs than the max supported. I also changed the collection to a `LinkedHashMap` to preserve the paths ordering.